kitsune/wiki/models.py

```diff
@@ -263,39 +279,53 @@ def save(self, *args, **kwargs):
+            title_changed and
+            # If the title was a template before, and is not now, or vice-versa
+            (self.title.startswith(TEMPLATE_TITLE_PREFIX) !=
+             self.old_title.startswith(TEMPLATE_TITLE_PREFIX)))
+
+        if slug_changed:
+            # Clear out the share link so it gets regenerated.
+            self.share_link = ''
 
         super(Document, self).save(*args, **kwargs)
 
         # Make redirects if there's an approved revision and title or slug
         # changed. Allowing redirects for unapproved docs would (1) be of
         # limited use and (2) require making Revision.creator nullable.
-        if self.current_revision and (slug_changed or title_changed):
+        #
+        # If the "templateness" changed, don't try and make a redirect.
+        # The category and the title-prefix have to stay consistent, and
+        # redirects are in the same category as their targets. There is not
+        # a clear way to construct the redirect without invalidating one of
+        # those conditions.
+        if self.current_revision and (slug_changed or title_changed) and not templateness_changed:
             try:
-                doc = Document.objects.create(locale=self.locale,
-                                              title=self._attr_for_redirect(
-                                                  'title', REDIRECT_TITLE),
-                                              slug=self._attr_for_redirect(
-                                                  'slug', REDIRECT_SLUG),
-                                              category=self.category,
-                                              is_localizable=False)
-                Revision.objects.create(document=doc,
-                                        content=REDIRECT_CONTENT % self.title,
-                                        is_approved=True,
-                                        reviewer=self.current_revision.creator,
-                                        creator=self.current_revision.creator)
+                doc = Document.objects.create(
+                    locale=self.locale,
+                    title=self._attr_for_redirect('title', REDIRECT_TITLE),
+                    slug=self._attr_for_redirect('slug', REDIRECT_SLUG),
+                    category=self.category,
+                    is_localizable=False)
+                Revision.objects.create(
+                    document=doc,
+                    content=REDIRECT_CONTENT % self.title,
+                    is_approved=True,
+                    reviewer=self.current_revision.creator,
+                    creator=self.current_revision.creator)
             except TitleCollision:
                 pass
 
-            if slug_changed:
-                del self.old_slug
-            if title_changed:
-                del self.old_title
+        if slug_changed:
+            del self.old_slug
+        if title_changed:
+            del self.old_title
```

Comment:
Awesome!
On Jul 23, 2015 3:57 PM, "Mike Cooper" <notifications@github.com> wrote:

> In kitsune/wiki/models.py
> <https://github.com/mozilla/kitsune/pull/2547#discussion_r35364175>:
>
> >              except TitleCollision:
> >                  pass
> >
> > -            if slug_changed:
> > -                del self.old_slug
> > -            if title_changed:
> > -                del self.old_title
> > +        if slug_changed:
> > +            del self.old_slug
> > +        if title_changed:
> > +            del self.old_title
>
> If a document became a template through a title change, it would have a
> self.old_title that needs cleaned up. But because of the not
> templateness_changed check in the previous if block, there would be cases
> where these attributes were created but not cleaned up. Dedenting them
> makes sure they always get cleaned up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/kitsune/pull/2547/files#r35364175>.
>